hnn_core/extracellular.py

```diff
@@ -0,0 +1,492 @@
+            H = np.dot(b, a) / norm_a  # NB can be negative
+            L = H + norm_a
+            R2 = np.dot(b, b) - H ** 2  # NB squares
+
+            # To avoid very large values when electrode is placed (anywhere) on
+            # the section axis, enforce minimal perpendicular distance
+            R2 = np.maximum(R2, (section.diam / 2.0) ** 2)
+
+            if L < 0 and H < 0:  # electrode is "behind" line segment
+                num = np.sqrt(H ** 2 + R2) - H  # == norm(b) - H
+                denom = np.sqrt(L ** 2 + R2) - L
+            elif L > 0 and H < 0:  # electrode is "on top of" line segment
+                num = (np.sqrt(H ** 2 + R2) - H) * (L + np.sqrt(L ** 2 + R2))
+                denom = R2
+            else:  # electrode is "ahead of" line segment
+                num = np.sqrt(L ** 2 + R2) + L
+                denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+            phi[idx] = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class ExtracellularArray:
+    """Class for recording extracellular potential fields with electrode array
+
+    Note that to add an electrode array to a simulation, you should use the
+    :meth":`hnn_core.Network.add_electrode_array`-method. After simulation,
+    the network will contain a dictionary of `ExtracellularArray`-objects
+    in `net.rec_array` (each array must be added with a unique name). The
+    `ExtracellularArray`s contain the voltages at each electrode contact,
+    along with the time points at which the voltages were sampled.
+
+    Parameters
+    ----------
+    positions : tuple | list of tuple
+        The (x, y, z) coordinates (in um) of the extracellular electrodes.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    method : str
+        Approximation to use. 'psa' (default) assigns all transmembrane
+        currents to the center point (0.5) (point source approximation).
+        'lsa' treats the section as a line source, and a single multiplier is
+        calculated for each section (line source approximation).
+
+    Attributes
+    ----------
+    times : list of float
+        The time points the extracellular voltages are sampled at (ms)
+    sfreq : float
+        Sampling rate of the extracellular data (Hz).
+
+    Notes
+    -----
+    See Table 5 in http://jn.physiology.org/content/104/6/3388.long for
+    measured values of sigma in rat cortex (note units there are mS/cm)
+    """
+
+    def __init__(self, positions, sigma=0.3, method='psa', times=list(),
+                 voltages=list()):
```

Comment:
this can be super surprising in Python. Never assign a mutable as a default value

Reply:
`None` better them?